kinesis/src/main/scala/com.snowplowanalytics.snowplow.collectors.scalastream/sinks/KinesisSink.scala

```diff
@@ -352,14 +349,60 @@ class KinesisSink private (
     executorService.awaitTermination(10000, MILLISECONDS)
     ()
   }
+
+  private def checkKinesisHealth(): Unit = {
+    val healthThread = new Thread() {
+      override def run() {
+        while (!kinesisHealhy) {
+          Try {
+            val describeRequest = new DescribeStreamSummaryRequest()
+            describeRequest.setStreamName(streamName)
+            val describeResult = client.describeStreamSummary(describeRequest)
+            describeResult.getStreamDescriptionSummary().getStreamStatus()
+          } match {
+            case Success("ACTIVE") =>
+              log.info(s"Stream $streamName ACTIVE")
+              kinesisHealhy = true
+            case Success(other) =>
+              log.warn(s"Stream $streamName not ACTIVE but $other")
+              Thread.sleep(1000L)
+            case Failure(err) =>
+              log.error(s"Error while checking status of stream $streamName: ${err.getMessage()}")
+              Thread.sleep(1000L)
```

Comment:
This is deliberate. The quotas for `describeStream` are per second and thus if we want to maximize the number of successful replies that we get we should be doing the calls once every second (unless I'm missing something).

Let's imagine that we have 100 collector instances starting at once. The first second, 20 instances will get a successful response, and the 80 other ones a quota exceeded exception. The 2nd second, 80 requests will be made again, 20 successful and 60 failures that will be retried the next second, etc. It's impossible to be quicker than this.

On top of that, all the collector instances are configured in the same way. Even if we use a different value, there are good chances that they will end up doing the requests at the same time.

Just as a reminder, this runs in the background and is not blocking, and if events are sent to Kinesis/SQS before the check succeeds, the sink is marked as healthy and the requests are stopped.

Reply:
And while all that is happening, any other service (non-collector) running within the account is unable to call `describeStream`, because the collector has stole all the quota.  For example, enrich-kinesis will give an error [on this line](https://github.com/snowplow/enrich/blob/3.7.0/modules/kinesis/src/main/scala/com/snowplowanalytics/snowplow/enrich/kinesis/Sink.scala#L90) and then it will exit with an error.

I'm fine with your reasoning, as long as you've thought through how it will impact other services.  I just thought that by making it configurable, we can make the collector less aggressive at using up all the quota, and we might appreciate having that flexibility if we notice odd things happening elsewhere.

Reply:
> For example, enrich-kinesis will give an error [on this line](https://github.com/snowplow/enrich/blob/3.7.0/modules/kinesis/src/main/scala/com/snowplowanalytics/snowplow/enrich/kinesis/Sink.scala#L90) and then it will exit with an error.

Good point! I think that we should remove that from Enrich, because we know that our infra configures the streams correctly. I think that we should let Enrich try to write to the streams and fail if it doesn't succeed. I'll create the issue.

>  I just thought that by making it configurable, we can make the collector less aggressive at using up all the quota, and we might appreciate having that flexibility if we notice odd things happening elsewhere.

You're right, I'll make this configurable, it can only be a good thing.

Reply:
I've just remembered that collector and Enrich won't interfere because collector [uses describeStreamSummary](https://github.com/snowplow/stream-collector/blob/feature/sink_health/kinesis/src/main/scala/com.snowplowanalytics.snowplow.collectors.scalastream/sinks/KinesisSink.scala#L360) and enrich [uses describeStream](https://github.com/snowplow/enrich/blob/3.7.0/modules/kinesis/src/main/scala/com/snowplowanalytics/snowplow/enrich/kinesis/Sink.scala#L90) and [they have different limits](https://docs.aws.amazon.com/streams/latest/dev/service-sizes-and-limits.html).

I'll still make it configurable.